File: scripts/extract-mockup.mjs
import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { readPsd, initializeCanvas } from 'ag-psd';
import { createCanvas as makeCanvas, createImageData, Image } from 'canvas';
import sharp from 'sharp';

// Initialize ag-psd with proper node-canvas bridge
initializeCanvas(
  (w, h) => makeCanvas(Math.max(1, w), Math.max(1, h)),
  (imgData) => {
    // imgData is an ImageData-like { width, height, data }
    const c = makeCanvas(imgData.width, imgData.height);
    const ctx = c.getContext('2d');
    const iData = ctx.createImageData(imgData.width, imgData.height);
    iData.data.set(imgData.data);
    ctx.putImageData(iData, 0, 0);
    return c;
  },
  (w, h) => createImageData(Math.max(1, w), Math.max(1, h)),
);

const psdPath = 'Mockups/Front-Back-Man-T-shirt-Mockup/Front-Back-Man-T-shirt-Mockup.psd';
const outDir = 'apps/web/public/mockups';
mkdirSync(outDir, { recursive: true });

console.log('Reading PSD file with composite image...');
const buffer = readFileSync(psdPath);

// Read with composite image data (the flattened result)
const psd = readPsd(buffer, {
  skipLayerImageData: true,
  skipCompositeImageData: false,
  skipThumbnail: true,
});

console.log(`Canvas: ${psd.width} x ${psd.height}`);

if (psd.canvas) {
  console.log('Got composite canvas, extracting...');
  const pngBuf = psd.canvas.toBuffer('image/png');
  writeFileSync(`${outDir}/tshirt-composite.png`, pngBuf);
  console.log(`Wrote composite: ${pngBuf.length} bytes`);

  // The composite is 3000x2000 with front on left half, back on right half
  const halfW = Math.floor(psd.width / 2);

  // Extract front half (left side)
  await sharp(pngBuf)
    .extract({ left: 0, top: 0, width: halfW, height: psd.height })
    .resize(800, null, { fit: 'inside' })
    .png()
    .toFile(`${outDir}/tshirt-front.png`);
  console.log('Wrote tshirt-front.png');

  // Extract back half (right side)
  await sharp(pngBuf)
    .extract({ left: halfW, top: 0, width: halfW, height: psd.height })
    .resize(800, null, { fit: 'inside' })
    .png()
    .toFile(`${outDir}/tshirt-back.png`);
  console.log('Wrote tshirt-back.png');

  console.log('\nDone! Check apps/web/public/mockups/');
} else {
  console.log('No composite canvas found. Trying individual layers...');

  // Re-read with layer image data
  const psd2 = readPsd(buffer, {
    skipLayerImageData: false,
    skipCompositeImageData: true,
    skipThumbnail: true,
  });

  function extractLayer(layers, name) {
    for (const l of layers) {
      if (l.name === name && l.canvas) return l;
      if (l.children) {
        const found = extractLayer(l.children, name);
        if (found) return found;
      }
    }
    return null;
  }

  // Try to find and export any layers with canvas data
  function dumpLayers(layers, prefix = '') {
    for (const l of layers) {
      if (l.canvas) {
        const fn = `${outDir}/layer-${prefix}${l.name.replace(/[^a-zA-Z0-9]/g, '_')}.png`;
        const buf = l.canvas.toBuffer('image/png');
        writeFileSync(fn, buf);
        console.log(`Wrote: ${fn} (${l.canvas.width}x${l.canvas.height})`);
      }
      if (l.children) dumpLayers(l.children, `${prefix}${l.name}_`);
    }
  }

  dumpLayers(psd2.children);
}

console.log(`\nCanvas: ${psd.width} x ${psd.height}`);
console.log(`Children (top-level layers): ${psd.children?.length ?? 0}\n`);

function printLayers(layers, indent = 0) {
  if (!layers) return;
  for (const layer of layers) {
    const prefix = '  '.repeat(indent);
    const type = layer.children ? 'GROUP' : 'LAYER';
    const vis = layer.hidden ? '(hidden)' : '(visible)';
    const size = layer.canvas ? `${layer.canvas.width}x${layer.canvas.height}` : 'no-canvas';
    const blend = layer.blendMode || '';
    console.log(`${prefix}${type} "${layer.name}" ${vis} ${size} blend:${blend} opacity:${layer.opacity ?? 1}`);
    if (layer.children) {
      printLayers(layer.children, indent + 1);
    }
  }
}

printLayers(psd.children);

// Also try to get the composite (flattened) image
if (psd.canvas) {
  console.log(`\nComposite canvas: ${psd.canvas.width}x${psd.canvas.height}`);
  // Write composite as raw RGBA
  const { width, height, data } = psd.canvas;
  // We'll use sharp to convert
  console.log(`Composite data length: ${data.length} (expected ${width * height * 4})`);
}

File: scripts/extract-mockup.py
"""Extract front/back t-shirt mockups from the PSD file."""
import os
from psd_tools import PSDImage
from PIL import Image

psd_path = 'Mockups/Front-Back-Man-T-shirt-Mockup/Front-Back-Man-T-shirt-Mockup.psd'
out_dir = 'apps/web/public/mockups'
os.makedirs(out_dir, exist_ok=True)

print(f'Loading PSD: {psd_path}')
psd = PSDImage.open(psd_path)
print(f'PSD size: {psd.width} x {psd.height}')

# List all layers
def print_layers(layers, indent=0):
    for layer in layers:
        vis = 'visible' if layer.is_visible() else 'hidden'
        kind = layer.kind
        bbox = layer.bbox if hasattr(layer, 'bbox') else 'N/A'
        print(f"{'  ' * indent}{kind} '{layer.name}' ({vis}) bbox={bbox} opacity={layer.opacity}")
        if layer.is_group():
            print_layers(layer, indent + 1)

print('\nLayer tree:')
print_layers(psd)

# Composite (flatten) the full PSD
print('\nCompositing full PSD...')
composite = psd.composite()
print(f'Composite size: {composite.size}, mode: {composite.mode}')

# Save full composite
composite.save(f'{out_dir}/tshirt-composite.png')
print(f'Saved tshirt-composite.png')

# The image is 3000x2000 with front on left half, back on right half
half_w = psd.width // 2

# Extract front (left half)
front = composite.crop((0, 0, half_w, psd.height))
# Resize to fit our 800x1000 canvas  
# Front half is 1500x2000, aspect ratio 0.75
# Our canvas is 800x1000, aspect ratio 0.8
# Let's resize to 800 wide, maintaining aspect
front_resized = front.resize((800, int(800 * psd.height / half_w)), Image.LANCZOS)
front_resized.save(f'{out_dir}/tshirt-front.png')
print(f'Saved tshirt-front.png ({front_resized.size})')

# Extract back (right half)
back = composite.crop((half_w, 0, psd.width, psd.height))
back_resized = back.resize((800, int(800 * psd.height / half_w)), Image.LANCZOS)
back_resized.save(f'{out_dir}/tshirt-back.png')
print(f'Saved tshirt-back.png ({back_resized.size})')

# Also try compositing with individual groups hidden/visible for analysis
# Let's also extract just the shirt without the color adjustment layers
# to understand what we have
print('\n--- Individual group composites ---')
for layer in psd:
    if layer.is_group():
        try:
            img = layer.composite()
            if img:
                fname = f'{out_dir}/group-{layer.name.replace(" ", "_").replace("/","_")}.png'
                img.save(fname)
                print(f'Saved {fname} ({img.size})')
        except Exception as e:
            print(f'Could not composite group "{layer.name}": {e}')

print('\nDone!')
